cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/replication/PullFilter.java

```diff
@@ -96,9 +96,11 @@ public PullFilter(String filterName) {
      */
     public PullFilter(String filterName, Map<String, String> parameters) {
```

Comment:
This constructor needs to be deprecated and a new constructor `PullFilter(String filterName, Map<String, String> parameters)` needs to be added to allow users to create their own `PullFilter`s with complex arguments in the `POST` body

Reply:
@tomblench What do you think about change the signature using generics? If we define the constructor as ```PullFilter(String filterName, Map<String, ? extends Object> parameters)``` the API will be backward compatible whereas the internal parameters storage could be changed to a ```Map<String,Object>```

Reply:
I was surprised that this works but it looks a bit hacky. I would prefer the original approach of a deprecated constructor and a new one as it makes it clear that we have made a change to the API here.

Reply:
The problem is that we can not deprecate the current constructor ```PullFilter(String filterName, Map<String, String> parameters)``` and create a new one ```PullFilter(String filterName, Map<String, Object> parameters)``` as due to type erasure both constructors will have the same signature. 

The only option is to generalise the current one.

It's true that it looks a bit hacky, but I dont see other mechanism that allows a smooth API evolution in this case. Using Map<String, ?> is a way to generalise the current definition.

Reply:
Good point. It looks like you can just change this to `PullFilter(String filterName, Map<String, Object> parameters) ` and it will not break existing code (binary or source). My understanding is that this works because of the way type erasure is done.

Reply:
If we define the constructor as Map<String,Object> we will break the existing source code. We will break the compilation of the existing projects. 

As an example, we get this error compiling the tests:
```
ReplicatorIdTest.java:160: error: incompatible types: Map<String,String> cannot be converted to Map<String,Object>
```
I only see the use of generics as a solution for this

Reply:
OK, go with the generics